Providing Context to the Ansible Automation for a Policy Violation (#73)
open-cluster-management-io/enhancements: enhancements/sig-policy/72-content-to-ansilbe-job/README.md

> # Add an "everyEvent" Ansible Job Run Mode to the Policy addon

Reviewer: Could you please fix the title?

  1. `policy.policyStatus.details[n].templateMeta`.
  2. `policy.policyStatus.details[n].history[n].eventName`. Here we can consider only keep the latest policy violation history, 
  which are `policy.policyStatus.details[0].history[0].lastTimestamp` and `policy.policyStatus.details[0].history[0].message`

### User Stories

#### Story 1

As an OCM Policy addon user, I would like to get the additional status in an Ansible job's `extra_vars` field. For example,
regarding the CertificatePolicy policy, which checks for expiring certificates in secrets, I want to get the replicated policy 
status for each non-compliant cluster so I can know which secret expires on a specific cluster in the Anisble Job. Then, the 
Ansible Automation Platform can create a notice ticket and provide the violation details. 

### Implementation Details/Notes/Constraints [optional]

N/A

### Risks and Mitigation

Since `policy_violation_context` is the replicated policy status for all non-compliant managed clusters, the length of 
`policy_violation_context` may be oversized when the amount of managed clusters is large.

> A possible solution only passes the top `N` replicated policy status into `policy_violation_context`, `N` is configurable

Reviewer 2: I think we probably want to specify how we are going to select the "top" N policy statuses to pass into the context. I'm not sure how we would prioritize which replicated policies make it in, so I think if there are more than "N" policies we would probably want to not include the replicated policy statuses at all.

Reviewer 2: Also, would we configure this with an env variable on the propagator?

Author: > I think we probably want to specify how we are going to select the "top" N policy statuses to pass into the context. I'm not sure how we would prioritize which replicated policies make it in, so I think if there are more than "N" policies we would probably want to not include the replicated policy statuses at all.

It is another solution,  we can discuss it if we decide such configure is needed.

Author: > Also, would we configure this with an env variable on the propagator?

I prefer it to be a field of policy automation.

- [ ] Test plan is defined
- [ ] Graduation criteria for dev preview, tech preview, GA
- [ ] User-facing documentation is created in
      [website](https://github.com/open-cluster-management-io/open-cluster-management-io.github.io/)

## Summary

The Open Cluster Management (OCM) Policy add-on Ansible integration should pass the additional information 
into an Ansible job when triggering the creation. It means that the Ansible user can get more policy violation 
information and use them in the Ansible playbook.

## Motivation

Policy Automation uses the `policyRef` field to match the target policy and monitor its non-compliance status. 
A non-compliance policy can trigger its Policy Automation to create an Ansible job on the Ansible Tower, which 
launches a predefined Ansible playbook against an inventory of hosts. During this creation, we can pass the 
additional information to the Ansible job as `extra_vars`.

The current limitation is that we only pass the non-compliant cluster as `extra_vars`, which isn't enough to 
describe the policy violation status. The following fields could be added:
1. The violated policy name
2. The policy violation message

> 3. The replicated policy status for each non-compliant cluster

Author: It is a good suggestion. Maybe we can consider adding all policy annotations or a new special Ansible Job annotation. The latter solution makes things a little complex.

> 3. The replicated policy status for each non-compliant cluster

Reviewer 3: Per discussion, there is already a field for this in the PolicyAutomation, so my suggestion is unnecessary.

> 3. The replicated policy status for each non-compliant cluster

Reviewer 4: Knowing the namespace of the "related resource" and annotations from the related resource would be useful to target the right secops or app sre owner.  Unfortunately we don't have related resources for certificate expiration yet - would love to see that added :-)

non-compliant cluster into the Ansible Job.

### Non-Goals

- None at this time

## Proposal
If policy automation runs on the hub cluster, it should not get the propagated policy status from the managed cluster,
which requires Search or Observability dependency. The same design also applies to the hub of the cluster scenario.

### New Content Structure

Current `extra_vars` has a field `target_clusters` contains non-compliant cluster name list:
```json
{
  "target_clusters": [
    "cluster1", 
    "cluster2", 
    "cluster3",
  ]
}
```

The new content structure will pass more information:
```json
{
  "target_clusters": [
    "cluster1", 
    "cluster2", 
    "cluster3",
  ],
  "policy_name": "policy1",
  "violation_message": "NonCompliant; violation -  xxxxxx",
  "policy_violation_context" : {
    "cluster1": "raw replicated policy status1",
    "cluster2": "raw replicated policy status2",
    },
}
```

- `policy_name` is the name of the non-compliance policy that triggers the ansible job.
  `violation_message` is the latest policy violation message. 

  Suppose the policy automation runs on the hub cluster. In that case, `policy_name` and `violation_message` are from 

>   the roof policy. 

Reviewer 2: root policy

Author: THX

>   Suppose the policy automation runs on the managed cluster. In that case, `policy_name` and `violation_message` are from the propagated policy on the managed cluster.

Author: You are correct, I updated the purpose.

  The field path of `policy_name` is `policy.metadata.name`.
  The field path of `violation_message` is `policy.policyStatus.details[0].history.message[0]`

- `policy_violation_context` is a map. The keys are non-compliant cluster names, and the value is the raw replicated 
  policy status for each non-compliant cluster.

  Suppose the policy automation runs on the hub cluster. In that case, the policy automation looks at the `cluster namespaces` 

>   of the hub cluster and gets the replicated policies. Then it filters out the non-compliance replicated policies and puts the raw status into the `policy_violation_context` map.

Reviewer 2: I would like to keep the structure of the status there @JustinKuli . But I agree that "raw status" implies that it will be a string

>   of the hub cluster and gets the replicated policies. Then it filters out the non-compliance replicated policies and puts the raw status into the `policy_violation_context` map.

Author: It will keep its structure. But we can trim it as @willkutler suggestion.

  Suppose the policy automation runs on the managed cluster. In that case, the policy automation looks at the propagated policy on the managed cluster, then `policy_violation_context` would be a map that only contains the propagated policy status.

>   The field path of the raw status is `policy.policyStatus`.

Reviewer 2: Is there any reason for users to get the entire raw status? I was thinking we would at least cut out the `status.details[n].templateMeta` field, since that is only used by the controller to maintain prune behavior. We also probably don't need `eventName` in `history`.

N/A

### Risks and Mitigation

Since `policy_violation_context` is the raw replicated policy status for all non-compliant clusters, the length of 
`policy_violation_context` may be oversized when the amount of managed clusters is large.

A possible solution only passes the top `N` replicated policy status into `policy_violation_context`, `N` is configurable
and the default value is a relatively small number (10 or 20).

## Design Details

See the [Proposal](#proposal) section.

### Open Questions [optional]

N/A

### Test Plan

- Unit tests can be implemented to check the created Ansible job. For each scenario, the `extra_vars` contain the designed data.

### Graduation Criteria

1. The proposal is reviewed and accepted
2. Implementation is completed to support the primary goals
3. Test cases developed to demonstrate that the user stories are fulfilled

### Upgrade / Downgrade Strategy

There are no concerns with upgrades since this is not a breaking change.

Downgrading would only pass `target_clusters` to Ansible Job, which is an acceptable outcome.

### Version Skew Strategy

N/A

## Implementation History

N/A

## Drawbacks

> 1. This does not provide customized `policy_violation_context` for different policies.

Author: I mean different fields/information for ConfigurationPolicy, CertificatePolicy here.

additional information to the Ansible job as `extra_vars`.

The current limitation is that we only pass the non-compliant cluster as `extra_vars`, which isn't enough to 
describe the policy violation status. The following fields could be added:
1. The violated policy name
2. The policy violation message
3. The replicated policy status for each non-compliant cluster

### Goals

Adding the name of the violated policy, the violation message, and the replicated policy status for each 
non-compliant cluster into the Ansible Job.

### Non-Goals

- None at this time

## Proposal
Policy automation runs on the hub cluster, it should not get the propagated policy status from the managed cluster,
which requires Search or Observability dependency. The same design also applies to the hub of the cluster scenario.

### New Content Structure

Current `extra_vars` has a field `target_clusters` contains non-compliant cluster name list:
```json
{
  "target_clusters": [
    "cluster1", 
    "cluster2", 
    "cluster3",
  ]
}
```

The new content structure will pass more information:
```json
{
  "target_clusters": [
    "cluster1", 
    "cluster2", 
    "cluster3",
  ],

>   "policy_name": "policy1",

Reviewer: I'm thinking it might be useful if we also included the policy namespace as an extra var in case the same policy name is used in multiple places.

Reviewer 4: yes, the namespace is needed.  A nice to have would be the acm hub clustername: example `policy-grc-cp-dev-lq5qm.dev08.red-chesterfield.com`.  I use it in my examples but I would guess it may be available some other way.

  "violation_message": "NonCompliant; violation -  xxxxxx",
  "policy_violation_context" : {

>     "local-cluster": {

Reviewer: To be consistent with what's in `target_clusters`.
```suggestion
    "cluster1": {
```

      "compliant": "NonCompliant",
      "details": [
        {
          "compliant": "NonCompliant",
          "history": [
            {
              "lastTimestamp": "2022-09-08T16:03:23Z",
              "message": "NonCompliant; violation - pods not found: [testpod] in namespace default missing"
            },
          ],
        },
      ]
    },
    "cluster2": {...},
    },
}
```

- `policy_name` is the name of the non-compliance policy that triggers the ansible job.
  `violation_message` is the latest policy violation message. 
  They are from the root policy on the hub cluster. 
  The field path of `policy_name` is `policy.metadata.name`.
  The field path of `violation_message` is `policy.policyStatus.details[0].history[0].message`

- `policy_violation_context` is a map. The keys are non-compliant cluster names, and the value is the raw replicated 
  policy status for each non-compliant cluster.
  The policy automation looks at the `cluster namespaces` of the hub cluster and gets the replicated policies. Then it filters out the non-compliant replicated policies and puts the status into the `policy_violation_context` map.

>   Suppose the policy automation runs on the managed cluster. In that case, the policy automation looks at the propagated policy on the managed cluster, then `policy_violation_context` would be a map that only contains the propagated policy status.

Reviewer: Just to clarify since the PolicyAutomation runs on the hub.
```suggestion
  Suppose the policy automation runs for the managed cluster. In that case, the policy automation looks at the propagated policy of the managed cluster, then `policy_violation_context` would be a map that only contains the propagated policy status.
```

- [ ] Graduation criteria for dev preview, tech preview, GA
- [ ] User-facing documentation is created in
 [website](https://github.com/open-cluster-management-io/open-cluster-management-io.github.io/)

## Summary

The Open Cluster Management (OCM) Policy add-on Ansible integration should pass the additional information 
into an Ansible job when triggering the creation. It means that the Ansible user can get more policy violation 
information and use them in the Ansible playbook.

## Motivation

Policy Automation uses the `policyRef` field to match the target policy and monitor its non-compliance status. 
A non-compliance policy can trigger its Policy Automation to create an Ansible job on the Ansible Tower, which 
launches a predefined Ansible playbook against an inventory of hosts. During this creation, we can pass the 
additional information to the Ansible job as `extra_vars`.

The current limitation is that we only pass the non-compliant cluster as `extra_vars`, which isn't enough to 
describe the policy violation status. The following fields could be added:
1. The violated policy name
2. The namespace of the root policy
3. The hub cluster name
4. The policy violation message

> 5. The replicated policy status for each non-compliant cluster

Reviewer 5: Do you consider adding policysets ( the policy belongs which policysets) ?

Author: We can do this

### Goals

Adding the name of the violated policy, the root policy namespace, the hub cluster name, the violation message, and 

> the replicated policy status for each non-compliant cluster into the Ansible Job.

Reviewer 5: That means user can get above info automatically in the ansible job after the job runs completed. correct ?

Author: Yes, in the `extra_vars` filed of the job on the ansible tower

### Non-Goals

- None at this time

## Proposal
Policy automation runs on the hub cluster. It should not get the propagated policy status from the managed cluster,

> which requires Search or Observability dependency. The same design also applies to the hub of the cluster scenario.

Reviewer 5: What does Observability have to do here ?

Author: I think it is mainly dependent on `Search` here, it can be updated if you think the current proposal isn't accurate.

Reviewer 5: Yes. As I known,  view job link will be switched from policy page to search page, it won't related with Observability-grafana.

### New Content Structure

Current `extra_vars` has a field `target_clusters` contains non-compliant cluster name list:
```json
{
  "target_clusters": [
    "cluster1", 
    "cluster2", 
    "cluster3",
  ]
}
```

The new content structure will pass more information:
```json
{
  "target_clusters": [
  "cluster1", 
  "cluster2", 
  "cluster3",
  ],
  "policy_name": "policy1",
  "namespace": "policy-namespace",

>   "hubcluster": "acm-hub-cluster",

Reviewer: ```suggestion
  "hub_cluster": "acm-hub-cluster",
```

            "lastTimestamp": "2022-09-08T16:03:23Z",
            "message": "NonCompliant; violation - pods not found: [testpod] in namespace default missing"
            },
          ],
        },
      ]
    },
    "cluster2": {
      ...
    },
  },
}
```

- `policy_name` is the name of the non-compliance policy that triggers the ansible job.
 `violation_message` is the latest policy violation message. 
 They are from the root policy on the hub cluster. 
 The field path of `policy_name` is `policy.metadata.name`.
 The field path of `violation_message` is `policy.policyStatus.details[0].history[0].message`

- `namespace` is the namespace of the root policy, which is from `rootPolicy.GetNamespace()`

- `hubcluster` is the name of the hub cluster, which is from `rootPolicy.GetClusterName()`

- `policy_violation_context` is a map. The keys are non-compliant cluster names, and the value is the raw replicated 
 policy status for each non-compliant cluster.
 The policy automation looks at the `cluster namespaces` of the hub cluster and gets the replicated policies. 
 Then it filters out the non-compliant replicated policies and puts the status into the `policy_violation_context` map.

>  Suppose the policy automation runs on the hub cluster without the managed cluster. In that case, `policy_violation_context` 

Reviewer: When would this happen? When running the automation manually?

Reviewer: In that case, wouldn't the entry be `local-cluster` and not be empty?

 1. `policy.policyStatus.details[n].templateMeta`.
 2. `policy.policyStatus.details[n].history[n].eventName`

### User Stories

#### Story 1

As an OCM Policy add-on user, I would like to get the additional status in an Ansible job's `extra_vars` field. For example,
regarding the CertificatePolicy policy, which checks for expiring certificates in secrets, I want to get the replicated policy 
status for each non-compliant cluster so I can know which secret expires on a specific cluster in the Anisble Job. Then, the 
Ansible Automation Platform can create a notice ticket and provide the violation details. 

### Implementation Details/Notes/Constraints [optional]

N/A

### Risks and Mitigation

Since `policy_violation_context` is the replicated policy status for all non-compliant managed clusters, the length of 
`policy_violation_context` may be oversized when the amount of managed clusters is large.

To solve this, we will only pass the top `N` replicated policy status into `policy_violation_context`, where `N` is configurable
via an optional policy automation field called `policyViolationContextLimit`. If this field is not set, `N` will default to the

> first 1000 managed clusters.

Reviewer 5: In large clusters, will the `policy_violation_context` impact performance when N is 1000 ? e.g  etcd, cpu

Author: We haven't tested which `N` will cause the performance issue, we can adjust `policyViolationContextLimit` in futrue

## Design Details

See the [Proposal](#proposal) section.

### Open Questions [optional]

N/A

### Test Plan

- Unit tests can be implemented to check the created Ansible job. For each scenario, the `extra_vars` contain the designed data.

### Graduation Criteria

1. The proposal is reviewed and accepted
2. Implementation is completed to support the primary goals
3. Test cases developed to demonstrate that the user stories are fulfilled

### Upgrade / Downgrade Strategy

There are no concerns with upgrades since this is not a breaking change.

Downgrading would only pass `target_clusters` to Ansible Job, which is an acceptable outcome.

### Version Skew Strategy

N/A

## Implementation History

N/A

## Drawbacks

> 1. This design does not restructure `policy_violation_context` data for different policies.

Reviewer 5: what's the "different policies" ?  different type policies ?  configuration and cert in one policy ?

Author: For different types of policies, we can parse the violation message from each replicated policy status. 

For example, for certificate policy, we can create the below object based on the parsed violation messages. This object is easier to use on the Ansible Tower. In the current scope, we decide to just pass the raw replicated policy status rather than parsing the violation messages.

{
    "cluster1": {
        "CertificatePolicy/certificate-policy-name": {
            "noncompliant_objects": [
                {
                    "name": "secret1",
                    "namespace": "my-project-namespace",
                    "kind": "Secret"
                },
                {
                    "name": "secret2",
                    "namespace": "my-project-namespace",
                    "kind": "Secret"
                },
                {
                    "name": "super-secret",
                    "namespace": "my-other-namespace",
                    "kind": "Secret"
                }
            ]
        }
    }
}
